packages/@webex/plugin-meetings/test/unit/spec/media/MediaConnectionAwaiter.ts

```diff
@@ -0,0 +1,326 @@
+import {assert} from '@webex/test-helper-chai';
+import sinon from 'sinon';
+import {ConnectionState, Event} from '@webex/internal-media-core';
+import testUtils from '../../../utils/testUtils';
+import {ICE_AND_DTLS_CONNECTION_TIMEOUT} from '@webex/plugin-meetings/src/constants';
+import MediaConnectionAwaiter from '../../../../src/media/MediaConnectionAwaiter';
+
+describe('MediaConnectionAwaiter', () => {
+  let mediaConnectionAwaiter;
+  let mockMC;
+  let clock;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers();
+
+    mockMC = {
+      getStats: sinon.stub().resolves([]),
+      on: sinon.stub(),
+      off: sinon.stub(),
+      getConnectionState: sinon.stub().returns(ConnectionState.New),
+      getIceGatheringState: sinon.stub().returns('new'),
+    };
+
+    mediaConnectionAwaiter = new MediaConnectionAwaiter({
+      webrtcMediaConnection: mockMC,
+    });
+  });
+
+  afterEach(() => {
+    clock.restore();
+    sinon.restore();
+  });
+
+  describe('waitForMediaConnectionConnected', () => {
+    it('resolves immediately if connection state is connected', async () => {
+      mockMC.getConnectionState.returns(ConnectionState.Connected);
+
+      await mediaConnectionAwaiter.waitForMediaConnectionConnected();
+
+      assert.neverCalledWith(mockMC.on);
+    });
+
+    it('rejects after timeout if ice state reach connected/completed', async () => {
+      mockMC.getConnectionState.returns(ConnectionState.Connecting);
+      mockMC.getIceGatheringState.returns('gathering');
+
+      let promiseResolved = false;
+      let promiseRejected = false;
+
+      mediaConnectionAwaiter
+        .waitForMediaConnectionConnected()
+        .then(() => {
+          promiseResolved = true;
+        })
+        .catch(() => {
+          promiseRejected = true;
+        });
+
+      assert.equal(promiseResolved, false);
+      assert.equal(promiseRejected, false);
+
+      // check the right listener was registered
+      assert.calledTwice(mockMC.on);
+      assert.equal(mockMC.on.getCall(0).args[0], Event.CONNECTION_STATE_CHANGED);
+      assert.equal(mockMC.on.getCall(1).args[0], Event.ICE_GATHERING_STATE_CHANGED);
+      const listener = mockMC.on.getCall(1).args[1];
+
+      mockMC.getIceGatheringState.returns('complete');
+      listener();
+
+      await clock.tickAsync(ICE_AND_DTLS_CONNECTION_TIMEOUT);
+      await testUtils.flushPromises();
+
+      assert.equal(promiseResolved, false);
+      assert.equal(promiseRejected, true);
+
+      assert.calledTwice(mockMC.off);
+    });
+
+    it('resolves after timeout if connection state reach connected/completed', async () => {
```

Comment:
this test case is only theoretical as in practice if the state changes to "connected" we should get a connection state change callback immediately and not have to wait for the timeout, maybe worth adding a comment saying that just so no one is confused in the future?

Reply:
It's not only theoretical as I have seen cases in production that suffered from missing this check before. I can share with you sample calls which encountered it.

Reply:
so in these cases you've seen we didn't get the callback about connection state change? if you have a log, I would be interested in seeing it,